### PR TITLE
feat: assumption sampler created with loss analysis feature

### PR DESCRIPTION
**Pull request by TheLemonPig**
## Description

Creates a new experimentalist that combats assumptions. Currently only and minimally implemented on loss. Samples are chosen based on which have the greatest increase in error when changing the loss function.

### Type of change:
- New feature (non-breaking change which adds functionality)

### Features: 
- Experimentalist/sampler

### Questions: 
- how do we want to pass Theorist information?

### Remarks: 
- This experimentalist requires experimental results and some theorist hyper-parameters (currently only what loss function it tried to minimize)

### CODE REVIEW

File: autora/experimentalist/sampler/assumption.py [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+import numpy as np
+from sklearn.metrics import mean_absolute_error as mae
+from sklearn.metrics import mean_squared_error as mse
+
+
+def assumption_sampler(X, model, n, y=None, loss=True, theory=None):
+    """
+
+    Args:
+        X: pool of IV conditions to evaluate uncertainty
```
Comment (musslick) [inline autora/experimentalist/sampler/assumption.py:12]:
```suggestion
        X: pool of IV conditions to sample from
```

Comment (musslick) [reply autora/experimentalist/sampler/assumption.py:12]: The doc string may not be up to date.

File: autora/experimentalist/sampler/assumption.py [hunk 2]
```diff
@@ -0,0 +1,50 @@
+from typing import Iterable
+
+import numpy as np
+from sklearn.metrics import mean_absolute_error as mae
+from sklearn.metrics import mean_squared_error as mse
+
+
+def assumption_sampler(X, model, n, y=None, loss=True, theory=None):
+    """
+
```
Comment (musslick) [inline autora/experimentalist/sampler/assumption.py:10]: You might also want to add a brief description of the sampler here

**Review by musslick — CHANGES_REQUESTED** [review]
This is neat! Before merging, you might want to add one or two unit tests that check the basic functionality.

File: tests/test_experimentalist_assumption.py [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+    """
+    # Make the data
+    def ground_truth(xs):
+        return (xs**2.0) + xs + 1.0
+
+    # X = np.random.randint(low=0, high=10, size=100).reshape(-1, 1)
+    X = np.array(range(11)).reshape(-1, 1)
+    y = np.array([x for x in ground_truth(X)])
+
+    # Train logistic regression model
```
Comment (TheLemonPig) [inline tests/test_experimentalist_assumption.py:28]:
```suggestion
    # Train ridge regression model
```

File: tests/test_experimentalist_assumption.py [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+    # assert model_loss_function == mean_squared_error
+    # sklearn does not quite seem to have functionality to retrieve loss function used by model
+    # you can get the scorer which corresponds to the loss function
+    # but there does not seem to be any built-in dictionary to get the corresponding function
+    model_loss_function = mean_squared_error
+    model_loss = model_loss_function(
+        y_true=y.reshape(1, -1), y_pred=y_pred.reshape(1, -1), multioutput="raw_values"
+    )
+    alternative_model = RidgeCV(scoring="neg_mean_absolute_error")
+    alternative_model.fit(X, y)
```
Comment (TheLemonPig) [inline tests/test_experimentalist_assumption.py:67]:
```suggestion
```

File: tests/test_experimentalist_assumption.py [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+    # alternative_model_loss_function =
+    # get_scorer(scoring=alternative_model.get_params()['scoring'])
```
Comment (TheLemonPig) [inline tests/test_experimentalist_assumption.py:69]:
```suggestion
    # calculate model loss on an alternative loss function
```

File: autora/experimentalist/sampler/assumption.py [hunk 6]
```diff
@@ -0,0 +1,69 @@
...
+        if loss:
+            if current is None:
+                current = mse
+                print(
+                    Warning(
+                        "Knowledge of Theorist Loss Function needed. MSE has been assumed."
+                    )
+                )
+            y_pred = model.predict(X)
+            print(y_pred.shape)
```
Comment (musslick) [inline autora/experimentalist/sampler/assumption.py:45]: We might want to avoid any print statements when regularly executing the code. Alternatively, you could include an additional boolean argument `verbose` in the function and then only print if verbose it set to true (the default would be false). E.g.,
```
if verbose:
   print(y_pred.shape)
```

**Review by musslick — CHANGES_REQUESTED** [review]
The unit test is a really nice addition. I only have a minor suggestion pertaining to the print statements; otherwise looks great!

File: autora/experimentalist/sampler/assumption.py (hunk 6 above)
Comment (TheLemonPig) [inline autora/experimentalist/sampler/assumption.py:45]:
```suggestion
```

Comment (TheLemonPig) [reply autora/experimentalist/sampler/assumption.py]: whoops! I didn't want that to be there at all. Thank you for catching that! I completely agree, we don't want print statements coming up normally. Since I deleted most of the print statements I don't see much use in keeping just this one in so I just removed it.

File: autora/experimentalist/sampler/assumption.py [hunk 7]
```diff
@@ -0,0 +1,68 @@
...
+        if loss:
+            if current is None:
+                current = mse
+                print(
+                    Warning(
+                        "Knowledge of Theorist Loss Function needed. MSE has been assumed."
+                    )
+                )
+            y_pred = model.predict(X)
+            print(y.shape)
```
Comment (musslick) [inline autora/experimentalist/sampler/assumption.py:45]: print statement here could be removed or be optional (in verbose mode)

File: autora/experimentalist/sampler/assumption.py [hunk 8]
```diff
@@ -0,0 +1,68 @@
...
+            print(X.shape)
```
Comment (musslick) [inline autora/experimentalist/sampler/assumption.py:46]: print statement here could be removed or be optional (in verbose mode)

File: autora/experimentalist/sampler/assumption.py [hunk 9]
```diff
@@ -0,0 +1,68 @@
...
+            )
+            print(current_loss)
+            alternative = mae
+            alternative_loss = alternative(
+                y_true=y.reshape(1, -1),
+                y_pred=y_pred.reshape(1, -1),
+                multioutput="raw_values",
+            )
+            loss_delta = alternative_loss - current_loss
+            print(loss_delta)
```
Comment (musslick) [inline autora/experimentalist/sampler/assumption.py:60]: print statement here could be removed or be optional (in verbose mode)

File: autora/experimentalist/sampler/assumption.py [hunk 10]
```diff
@@ -0,0 +1,68 @@
...
+            idx = np.flip(loss_delta.argsort()[:n])
+            print(idx)
```
Comment (musslick) [inline autora/experimentalist/sampler/assumption.py:62]: print statement here could be removed or be optional (in verbose mode)

**Review by musslick — CHANGES_REQUESTED** [review]
Looking great! There are still a few print statements that could either be removed or be called in verbose mode...

File: autora/experimentalist/sampler/assumption.py (hunk 7 above)
Comment (TheLemonPig) [inline autora/experimentalist/sampler/assumption.py:45]:
```suggestion
```

File: autora/experimentalist/sampler/assumption.py (hunk 8 above)
Comment (TheLemonPig) [inline autora/experimentalist/sampler/assumption.py:46]:
```suggestion
```

File: autora/experimentalist/sampler/assumption.py (hunk 9 above)
Comment (TheLemonPig) [inline autora/experimentalist/sampler/assumption.py:60]:
```suggestion
```

File: autora/experimentalist/sampler/assumption.py (hunk 10 above)
Comment (TheLemonPig) [inline autora/experimentalist/sampler/assumption.py:62]:
```suggestion
```

**Comment by TheLemonPig** [comment]
> Looking great! There are still a few print statements that could either be removed or be called in verbose mode...

Sorry about that! All is dealt with now

File: autora/experimentalist/sampler/assumption.py [hunk 11]
```diff
@@ -0,0 +1,64 @@
+from typing import Iterable
+
+import numpy as np
+from sklearn.metrics import mean_absolute_error as mae
+from sklearn.metrics import mean_squared_error as mse
+
+
+def assumption_sampler(X, y, model, n, loss=True, theorist=None):
```
Comment (TheLemonPig) [inline autora/experimentalist/sampler/assumption.py:8]:
```suggestion
def assumption_sampler(X, y, model, n, loss=True, theorist=None, confirmation_bias=False):
```

Comment (TheLemonPig) [reply autora/experimentalist/sampler/assumption.py:8]: added parameter

File: autora/experimentalist/sampler/assumption.py [hunk 12]
```diff
@@ -0,0 +1,64 @@
...
+            )
+            print(current_loss)
+            alternative = mae
+            alternative_loss = alternative(
+                y_true=y.reshape(1, -1),
+                y_pred=y_pred.reshape(1, -1),
+                multioutput="raw_values",
+            )
+            loss_delta = alternative_loss - current_loss
+            idx = np.flip(loss_delta.argsort()[:n])
```
[the remark this answers is not in the document]
Comment (TheLemonPig) [reply autora/experimentalist/sampler/assumption.py:58]: added feature linked to added parameter

File: autora/experimentalist/sampler/assumption.py [hunk 13]
```diff
@@ -0,0 +1,64 @@
...
+    Because it samples with respect to a Theorist, this sampler cannot be used on the first cycle
+
+    Args:
+        X: pool of IV conditions to sample from
+        y: experimental results from most recent iteration
+        model: Scikit-learn model, must have `predict` method.
+        n: number of samples to select
+        loss: assumption to test: identify points that are most affected by choice of loss function
+        theorist: the Theorist, which employs the theory it has been hard-coded to demonstrate
+
```
Comment (TheLemonPig) [inline autora/experimentalist/sampler/assumption.py:23]:
```suggestion
        confirmation_bias: find evidence to support or undermine the theory

```

File: autora/experimentalist/sampler/assumption.py [hunk 14]
```diff
@@ -0,0 +1,64 @@
...
+    sub-algorithms e.g. loss function
+    Because it samples with respect to a Theorist, this sampler cannot be used on the first cycle
+
+    Args:
+        X: pool of IV conditions to sample from
+        y: experimental results from most recent iteration
+        model: Scikit-learn model, must have `predict` method.
+        n: number of samples to select
+        loss: assumption to test: identify points that are most affected by choice of loss function
+        theorist: the Theorist, which employs the theory it has been hard-coded to demonstrate
```
Comment (TheLemonPig) [inline autora/experimentalist/sampler/assumption.py:22]:
```suggestion
        theorist: the Theorist, which employs the theory it has been hard-coded to demonstrate
        confirmation_bias: whether to find evidence to support or oppose the theory
```

**Review by musslick — APPROVED** [review]
Looking great!! Thanks a lot for adding that. We might just want to confirm with @hollandjg about the online checks not working. I don't think it has to do with the code but perhaps with the poetry setup...

**Comment by gtdang** [comment]
> Looking great!! Thanks a lot for adding that. We might just want to confirm with @hollandjg about the online checks not working. I don't think it has to do with the code but perhaps with the poetry setup...

It seems like it's failing on the black pre-commit hook. Could you try running black on the code to reformat and committing those changes?

**Comment by TheLemonPig** [comment]
@hollandjg @gtdang do we know why these tests might be failing?

**Comment by gtdang** [comment]
> @hollandjg @gtdang do we know why these tests might be failing?

The error message says that the grid_pooler cannot be imported.
Screen Shot 2023-01-09 at 9 41 09 AM

The path to this function has changed in main. 
Screen Shot 2023-01-09 at 9 44 40 AM

Try fetching and updating your local autora and change the test script with the import statement:
`from aurora.experimentalist.pooler import grid_pool`
